torch/csrc/distributed/rpc/rref.cpp

```diff
@@ -131,32 +130,39 @@ const ForkId& UserRRef<T>::forkId() const {
 }
 
 template <>
-IValue UserRRef<IValue>::toHere() {
+std::shared_ptr<ivalue::Future> UserRRef<IValue>::toHere() {
+  auto future = std::make_shared<ivalue::Future>(nullptr);
   auto agent = RpcAgent::getDefaultRpcAgent();
-  std::shared_ptr<FutureMessage> fm = agent->send(
+  auto fm = agent->send(
       agent->getWorkerInfo(ownerId_),
       ScriptRRefFetchCall(rrefId()).toMessage());
-  const Message& message = fm->wait();
-  RRefContext::handleException(message);
-  auto rfr = RRefFetchRet::fromMessage(message);
-  TORCH_INTERNAL_ASSERT(
-      rfr->values().size() == 1,
-      "RRef of IValue should contain a single IValue, but got ",
-      rfr->values().size());
-  return rfr->values().front();
+
+  fm->addCallback([future](const Message& message) {
+    RRefContext::handleException(message);
+    auto rfr = ScriptRRefFetchRet::fromMessage(message);
```

Comment:
Actually, I am not sure if we would like to do that. `deserializeResponse` uses switch case on the message type to find the correct way to deserialize the response, but here we know for sure that the response should be  `ScriptRRefFetchRet` otherwise sth is wrong. Will it be reasonable if we directly call `ScriptRRefFetchRet::fromMessage` here to save the overhead of switch case?

Reply:
Makes sense. They are in pair.

Then, why do we need to add `SCRIPT_RREF_FETCH_RET` and `PYTHON_RREF_FETCH_RET` branches to `deserializeResponse `. Where is it used?

Reply:
Good catch!! They don't need to be in `deserializeResponse`. Let me remove them.